edgedb/protocol/codecs/codecs.pyx

```diff
@@ -381,6 +379,31 @@ cdef duration_decode(pgproto.CodecContext settings, FRBuffer *buf):
     return datatypes.new_duration(microseconds, days, months)
 
 
+cdef decimal_decode(pgproto.CodecContext settings, FRBuffer *buf):
+    return pgproto.numeric_decode_binary_ex(settings, buf, True)
+
+
+cdef bigint_encode(pgproto.CodecContext settings, WriteBuffer buf, obj):
+    if isinstance(obj, float):
+        raise TypeError(
+            f'expected an integer or an integral Decimal'
+        )
+    elif isinstance(obj, decimal.Decimal):
```

Comment:
I don't think we should have any association with decimal for std::bigint -- can't we just use Python's `int` type?

Reply:
I'd thought about that.  One complication is that we cannot distinguish regular ints from bigints in repl.  I can make a subclass of int to work this around, would that do?

Reply:
>  I can make a subclass of int to work this around, would that do?

Two ways to solve this:

* A subclass. We can try that now.

* Add more introspection utilities to edgedb-python (currently private API) to introspect the codec of any value being rendered.

What I'm certain about is that repl shouldn't guide this decision at all -- `std::bigint` should be a Python `int` as it's just a simpler type to work with for users.

Reply:
Why you need this in repl? Just for `123n` in repr? If just for that we might skip it for now. As its trivial in rust and we're going to make repl rustish in Alpha 3, I think.

Reply:
OK. I've made it decode as a regular `int`.  Accepting integral instances of `decimal.Decimal` is still fine, IMO.

Reply:
> OK. I've made it decode as a regular int. Accepting integral instances of decimal.Decimal is still fine, IMO.

Isn't it a lot like accepting integral floats? Isn't this a mis-feature that we add just because we can?

Reply:
No. `Decimal` instances with `exponent >= 0` can be considered bigints.  I.e we do not accept `Decimal('1.0'), but we _do_ accept `Decimal('12345')`. Whereas there is no such thing as integral float `1.0` is still a float.

The cost of supporting this is practically zero, and I don't think it's a misfeature.  That said, if you are strongly against, I won't be dying on this hill.

Reply:
> if you are strongly against, I won't be dying on this hill.

Let me put it this way: I'm not strongly against, but I don't understand why this is needed.

Especially in Python where the builtin `int` is already unlimited precision and people don't use decimals for this (even in Python 2 there was the `long` type).

Reply:
Ironically, in other int encoders, the check is *weaker*. `if type(obj) is not int and hasattr(type(obj), '__int__'):` happily allows `Decimal` instances, even those with `exponent < 0`:

```python
print(con.fetchone('SELECT <int64>$arg', arg=decimal.Decimal('10.10')))
```

This returns `10`.  We obviously need to fix this, but then to prohibit `decimal` we need to *actively* ban it, which, I think is a bit pointless.

Reply:
Actually, this is even more interesting. We have explicit casts from floats / decimals to ints in EdgeQL.  So question: why should the encoder be more strict than a cast?

Reply:
> We obviously need to fix this,

Yes

> We have explicit casts from floats / decimals to ints in EdgeQL. So question: why should the encoder be more strict than a cast?

Explicit, not implicit.  Allowing `fetchone('SELECT <int64>$1', 1.2)` would be similar to an implicit cast from float to int, something that we don't have in edgeql.

Reply:
Well, technically, that `<int64>` _is_ explicit, but I see the point.

Reply:
I look at casts for arguments as a hint for the argument's type, not actual casts. At least I remember us defining them that way. 

If/when in the future we have a better type analyses we should be able to infer parameter types in expressions like `1 + $a`, where `$a` would be an `int64`. If you relax the semantics now this feature will be extremely confusing, as the result of `fetchone(`SELECT 1 + $a`, a=1.2)` call would be `2`, not `2.2`. If we reject `1.2` now we won't have this problem.

Reply:
OK, I've made passing a non-int an error.